_posts/2024-10-31-WildFly-playing-with-generative-ai.adoc

```diff
@@ -0,0 +1,355 @@
+This embedding model will be used to compute the embedding of the user query and then the application will search for the nearest contents in the in-memory embedding store.
+This content retriever will append those contents to the user query to create the prompt that will be sent to the LLM we are connected to, via the ChatLanguageModel.
+
+==== Running Ollama locally
+
+[source, bash]
+----
+podman run -d --rm --name ollama --replace --pull=always -p 11434:11434 -v ollama:/root/.ollama --stop-signal=SIGKILL docker.io/ollama/ollama
+----
+
+Execute the following command to select the expected model (type */bye* to quit the ollama prompt):
+[source, bash]
+----
+podman exec -it ollama ollama run llama3.1:8b
+----
+
+==== Configuring the server
+
+We will use the following layers:
+
+* default-embedding-content-retriever
+* ollama-chat-model
+
+So the pom.xml should look like this:
+
+[source, xml]
+----
+<!-- The WildFly plugin deploys your war to a local JBoss AS container -->
+<plugin>
+    <groupId>org.wildfly.plugins</groupId>
+    <artifactId>wildfly-maven-plugin</artifactId>
+    <version>${version.wildfly.maven.plugin}</version>
+    <configuration>
+        <feature-packs>
+            <feature-pack>
+                <location>org.wildfly:wildfly-galleon-pack:${version.wildfly.bom}</location>
+            </feature-pack>
+            <feature-pack>
+                <location>org.wildfly:wildfly-ai-galleon-pack:0.1.0</location>
+            </feature-pack>
+        </feature-packs>
+        <layers>
+            <layer>cloud-server</layer>
+            <layer>default-embedding-content-retriever</layer>
+            <layer>ollama-chat-model</layer>
+        </layers>
+        <name>ROOT.war</name>
+        <packagingScripts>
+            <packaging-script>
+                <scripts>
+                    <script>./src/main/resources/scripts/configure_llm.cli</script>
+                </scripts>
+            </packaging-script>
+        </packagingScripts>
+    </configuration>
+    <executions>
+        <execution>
+            <goals>
+                <goal>package</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+
+==== The code
+
+The code itself is quite straightforward:
+[source, java]
+----
+@ServerEndpoint(value = "/websocket/chatbot",
+        configurator = org.wildfly.ai.websocket.CustomConfigurator.class)
+public class RagChatBot {
+    @Inject
+    @Named(value = "ollama")
+    ChatLanguageModel chatModel;
+    @Inject
+    @Named(value = "embedding-store-retriever")
+    ContentRetriever retriever;
+
+  private static final String PROMPT_TEMPLATE = "You are a WildFly expert who understands well how to administrate the WildFly server and its components\n"
+            + "Objective: answer the user question delimited by  ---\n"
+            + "\n"
+            + "---\n"
+            + "{{userMessage}}\n"
+            + "---"
+            + "\n Here is a few data to help you:\n"
```

Comment:
For "Here is a few data...", something like "Here is some data..." or "Here is some additional content..." sounds more correct